HIP/hip-980.md

```diff
@@ -0,0 +1,114 @@
+This change will be especially valuable for building the foundation of the so-called shadow fork functionality, where a user or developer can use a local mirror node instance but run it against a production environment state.
+Having this support, users or developers will have the ability to use public state data to simulate and troubleshoot existing issues or replay historical transactions using local mirror node instance against production state.
+
+## Rationale
+
+Users may want to run historical transactions on a local mirror node instance using public state data. To make this possible, some of the existing REST APIs should be enhanced with a timestamp query parameter, so that historical public data can be fetched and used
+for the simulations.
+
+## User stories
+
+- As a user of mirror node, I want to be able to query historical data for accounts, contracts, tokens, and transactions that were created at a specific timestamp.
+- As a user of mirror node, I want to be able to run locally an `eth_call`, `eth_estimateGas` or `eth_debugTraceTransaction` call that fails on a production environment against the corresponding public state.
+- As a developer of mirror node, I want to be able to test new feature locally against a real life data from a public environment, so that I have better understanding of the feature behavior.
+
+## Specification
+
+The proposed enhancement involves adding a query parameter `timestamp` to the existing routes:
+
+- `/api/v1/accounts/{idOrAliasOrEvmAddress}/tokens` to return correct historical information about an account and their associated fungible tokens for a specific timestamp
+- `/api/v1/accounts/{idOrAliasOrEvmAddress}/nfts` to return correct historical information about an account and their associated nfts for a specific timestamp
+- `/api/v1/accounts/{idOrAliasOrEvmAddress}/allowances/crypto` to return correct historical information about an account's crypto allowances for a specific timestamp
+- `/api/v1/accounts/{idOrAliasOrEvmAddress}/allowances/nfts` to return correct historical information about an account's NFT allowances for a specific timestamp
+- `/api/v1/accounts/{idOrAliasOrEvmAddress}/allowances/tokens` to return correct historical information about an account's token allowances for a specific timestamp
+- `/api/v1/tokens/{tokenId}/nfts/{serialNumber}` to return correct historical information about an NFT for a specific timestamp
+
+**Timestamp Query Parameter**
+
+A new `timestamp` query parameter will be added to search for different resources. This query parameter will represent a consensus timestamp in a Unix format in `seconds.nanoseconds` format and will have an optional comparison operator. That is, it will search for results that `equal`, `not equal`, are `greater` or `less` than the timestamp itself or its seconds component.
```

Comment:
Though it spells out the operators, it doesn't actually list the operator names like `eq`, `lt`, etc. We should also add an example below with operators

Reply:
Fixed